packages/Sandblocks-Core/SBGroup.class.st

```diff
@@ -0,0 +1,310 @@
+		ifFalse: [verticalEdges do: edgeBlock].
+]
+
+{ #category : #'as yet unclassified' }
+SBGroup >> addWithoutForce: aForceMoveDecorator [
+
+	self createMissingForcesFor: aForceMoveDecorator.
+	self blocks add: aForceMoveDecorator morph.
+	aForceMoveDecorator group: self.
+	aForceMoveDecorator morph when: #positionChanged send: #computeVertices to: self.
+	aForceMoveDecorator morph when: #extentChanged send: #computeVertices to: self.
+	self computeVertices.
+	aForceMoveDecorator morph
+		when: #delete
+		send: #remove:
+		to: self
+		with: aForceMoveDecorator morph.
+	aForceMoveDecorator morph
+		when: #spawn
+		send: #add:
+		to: self
+		with: aForceMoveDecorator.
+	self indicator
+		ifNil: [
+			self indicator: (SBOffscreenBlockIndicator newFor: self).
+			aForceMoveDecorator morph sandblockEditor openMorphInOverlay: self indicator]
+		ifNotNil: [self indicator updateLabel]
+]
+
+{ #category : #accessing }
+SBGroup >> blocks [
+
+	^ blocks
+]
+
+{ #category : #accessing }
+SBGroup >> blocks: aCollection [
+
+	blocks := aCollection.
+]
+
+{ #category : #'as yet unclassified' }
+SBGroup >> computeVerticalEdges [
+	
+	| verticalEdges |
+	verticalEdges := OrderedCollection new.
+	
+	self blocks do: [:aBlock |
+		verticalEdges add: (SBGroupBlockEdge leftEdgeOf: aBlock).
+		verticalEdges add: (SBGroupBlockEdge rightEdgeOf: aBlock)].
+	
+	verticalEdges sort: [:a :b | a position < b position].
+	
+	^ verticalEdges
+]
+
+{ #category : #'as yet unclassified' }
+SBGroup >> computeVertices [
+	
+	| verticalEdges newVertices |
+	verticalEdges := self computeVerticalEdges.	
+	newVertices := OrderedCollection new.
+	
+	self
+		addVerticesTo: newVertices from: verticalEdges side: #top;
+		addVerticesTo: newVertices from: verticalEdges side: #bottom;
+		setVertices: newVertices.
+]
+
+{ #category : #'as yet unclassified' }
+SBGroup >> createMissingForcesFor: aForceMoveDecorator [
+
+	| missingConnectionsTo |
+	missingConnectionsTo := self blocks copy.
+	aForceMoveDecorator additionalForces do: [:aForce |
+		missingConnectionsTo remove: aForce to ifAbsent: [].
+		missingConnectionsTo remove: aForce from ifAbsent: [].
+		
+		].
+	missingConnectionsTo do: [:aBlock |
+		aForceMoveDecorator morph sandblockEditor ifNotNil: [:editor |
+			editor openMorph: (SBForceConnector
+				connect: aForceMoveDecorator morph
+				to: aBlock
+				forceScale: 0.5)]]
+]
+
+{ #category : #'as yet unclassified' }
+SBGroup >> handlesMouseDown: evt [
+
+	^ true
+]
+
+{ #category : #'as yet unclassified' }
+SBGroup >> handlesMouseMove: anEvent [
+
+	^ true
+]
+
+{ #category : #'as yet unclassified' }
+SBGroup >> handlesMouseOverDragging: evt [
+
+	^ true
+
+]
+
+{ #category : #accessing }
+SBGroup >> indicator [
+
+	^ indicator
+]
+
+{ #category : #accessing }
+SBGroup >> indicator: aMorph [
+
+	indicator := aMorph.
+]
+
+{ #category : #'as yet unclassified' }
+SBGroup >> initialize [
+
+	super initialize.
+	self blocks: Set new.
+	self color: (Color random alpha: 0.5).
```

Comment:
what was your experience so far with the generated colors? If many of them ended up very bland you could consider doing something like `Color h: 360 atRandom s: 1 v: 1` instead (will of course cause more duplicates!)

Reply:
I think they seemed fine, we made them fit via the alpha value :D